pkg/ingester/label_names_and_values_test.go

```diff
@@ -77,207 +81,151 @@ func TestLabelNamesAndValuesAreSentInBatches(t *testing.T) {
 
-func TestLabelValues_CardinalityReportSentInBatches(t *testing.T) {
-	existingLabels := map[string][]string{
-		"lbl-a": {"a0000000", "a1111111", "a2222222"},
-		"lbl-b": {"b0000000", "b1111111", "b2222222", "b3333333"},
-		"lbl-c": {"c0000000"},
-		"lbl-d": {"d0000000"},
-		"lbl-e": {"e0000000"},
-		"lbl-f": {"f0000000", "f1111111", "f2222222"},
-		"lbl-g": {"g0000000"},
+	// First batch will contain exactly queryStreamBatchMessageSize bytes if label-a values
+	for i := 0; i < maxBatchLabelValues; i++ {
+		writeReq.Timeseries = append(writeReq.Timeseries, seriesForLabel("label-a"))
 	}
-	// server
-	mockServer := &mockLabelValuesCardinalityServer{context: context.Background()}
-	var server client.Ingester_LabelValuesCardinalityServer = mockServer
-
-	// index reader
-	idxReader := &mockIndex{existingLabels: existingLabels}
-	postingsForMatchersFn := func(reader tsdb.IndexPostingsReader, matcher ...*labels.Matcher) (index.Postings, error) {
-		return &mockPostings{n: 100}, nil
+
+	// Second batch will contain the last value of the label-a, and (maxBatchLabelValues-1) values of the label-b
+	writeReq.Timeseries = append(writeReq.Timeseries, seriesForLabel("label-a"))
+	for i := 0; i < maxBatchLabelValues-1; i++ {
+		writeReq.Timeseries = append(writeReq.Timeseries, seriesForLabel("label-b"))
 	}
-	err := labelValuesCardinality(
-		[]string{"lbl-a", "lbl-b", "lbl-c", "lbl-d", "lbl-e", "lbl-f", "lbl-g"},
-		[]*labels.Matcher{},
-		idxReader,
-		postingsForMatchersFn,
-		25,
-		server,
-	)
+
+	// Third batch will contain the last value of the label-b
+	writeReq.Timeseries = append(writeReq.Timeseries, seriesForLabel("label-b"))
+
+	// Write the series
+	_, err := in.Push(ctx, writeReq)
 	require.NoError(t, err)
 
-	require.Len(t, mockServer.SentResponses, 4)
+	// Query
+	mockServer := &mockLabelValuesCardinalityServer{context: ctx}
+	req := &client.LabelValuesCardinalityRequest{
+		LabelNames: []string{"label-a", "label-b"},
+		Matchers:   nil,
+	}
 
-	require.Equal(t, []*client.LabelValueSeriesCount{
-		{
-			LabelName: "lbl-a",
-			LabelValueSeries: map[string]uint64{
-				"a0000000": 100,
-				"a1111111": 100,
-				"a2222222": 100,
-			},
-		},
-		{
-			LabelName: "lbl-b",
-			LabelValueSeries: map[string]uint64{
-				"b0000000": 100,
-			},
-		},
-	}, mockServer.SentResponses[0].Items)
+	err = in.LabelValuesCardinality(req, mockServer)
+	require.NoError(t, err)
+	require.Len(t, mockServer.SentResponses, 3, "There should be three responses")
 
-	require.Equal(t, []*client.LabelValueSeriesCount{
-		{
-			LabelName: "lbl-b",
-			LabelValueSeries: map[string]uint64{
-				"b1111111": 100,
-				"b2222222": 100,
-				"b3333333": 100,
-			},
-		},
-		{
-			LabelName: "lbl-c",
-			LabelValueSeries: map[string]uint64{
-				"c0000000": 100,
-			},
-		},
-	}, mockServer.SentResponses[1].Items)
+	require.Equal(t, 1, len(mockServer.SentResponses[0].Items), "First response should contain only one label")
```

Comment:
I think there's a testify assertion for length `require.Len()` or similar.

If these assertions fail, it will be kinda hard to debug what's wrong since we're only comparing the length. Would it be possible to compare equality with some expected values? Or is the output too hard to predict?

Reply:
I didn't want to use `require.Len` as when it errors it prints a message including the elements of the slice, but here it's a one megabyte slice, so it would be useless.

OTOH, I tried asserting each message's specific label values, but they come unordered from the IndexReader, so there's no way to make it deterministic unless we sort them just for the test, which doesn't make sense IMO.